Add Lucene query syntax help

Bargs opened the pull request:
Fixes https://github.com/elastic/kibana/issues/4250

Initially I was going to create a separate dropdown for this help text but integrating it with the typeahead felt like a more elegant solution. The help text will appear when the query bar is empty and has focus. When the user starts typing the help text will go away and the usual typeahead suggestions appear. For now the help text is static, but I think this paves the way for more dynamic help similar to Slack's search box.

The help content still needs some design polish, after which I'll also add it to the Visualize and Dashboard query bars.

<img width="1680" alt="screen shot 2017-03-29 at 11 43 36 am" src="https://cloud.githubusercontent.com/assets/6239176/24463973/69745284-1476-11e7-84d8-0e427697f758.png">

epixa commented:
With this implementation, users won't be able to reference help if they've already started building the query.  I don't think that's a deal breaker upfront, but it's probably something we should address at some point.  This certainly still has value as-is, though.

Bargs commented:
@cjcenizal I could use some advice on how to best go about getting the design closer to the  [mockup you posted](https://github.com/elastic/kibana/issues/4250#issuecomment-280181432). I applied some of the kui styles you suggested but I think there's still some room for improvement.

Bargs commented:
@epixa my long term plan to address that is to make the help and suggestions more dynamic, similar to slack's search bar. In the short term I'm assuming if someone wants to copy one of the these static examples, they'll copy paste rather than try to type it in manually. That was just my line of thinking at least.

cjcenizal commented on src/ui/public/typeahead/partials/typeahead-items.html:
```diff
@@ -5,7 +5,10 @@
   class="typeahead-items"
   >
 
+  <div ng-if="!typeahead.query.length" ng-transclude></div>
```
If you remove this ng-if and put the transclude after the suggestions, then users can access the help even when the typeahead is offering suggestions. We might need to figure out a way to let the use collapse/expand the help. I think we could do that in a subsequent PR.

![image](https://cloud.githubusercontent.com/assets/1238659/24471861/57173bf4-1478-11e7-94a7-dda9980ba1d4.png)


```html
<div
  ng-show="typeahead.isVisible()"
  ng-mouseenter="typeahead.setMouseover(true);"
  ng-mouseleave="typeahead.setMouseover(false);"
  class="typeahead-items"
  >

  <div
    ng-if="typeahead.query.length"
    ng-repeat="item in typeahead.getItems()"
    ng-class="{active: item === typeahead.active}"
    ng-click="typeahead.selectItem(item, $event);"
    ng-mouseenter="typeahead.activateItem(item);"
    class="typeahead-item"
    >
    {{item}}
  </div>

  <div ng-transclude></div>
</div>
```

Bargs replied:
I played around with that idea before, I was afraid it might be too noisy. I'm down with it if you think it looks ok though.

cjcenizal replied:
I think it's better than hiding it. It's definitely not perfect though... could use some of @snide's magic.

Bargs commented:
@cjcenizal I updated the styles and changed the transclude to always be present below the typeahead items.

`kuiTable--fluid` looked a bit odd on wide screens so I tried removing it. Let me know what you think.

With `kuiTable--fluid`:

![screen shot 2017-04-03 at 12 07 56 pm](https://cloud.githubusercontent.com/assets/6239176/24619310/92917404-1867-11e7-977f-c64becd9e43d.png)

Without `kuiTable--fluid`:

![screen shot 2017-04-03 at 12 08 59 pm](https://cloud.githubusercontent.com/assets/6239176/24619321/9b25a3b0-1867-11e7-825a-1511911bc1fc.png)

cjcenizal commented:
@Bargs I chose `kuiTable--fluid` because without it the table will become less usable as screen-width grows. Here's how it looks on my 27":

![image](https://cloud.githubusercontent.com/assets/1238659/24619938/a5755cdc-1850-11e7-887e-e09887f39d8d.png)

cjcenizal commented:
@snide Thoughts on this? ^

snide commented:
@cjcenizal Yeah, I'd keep off the full width since that page itself is full width and it'll expand into unreadability.

Bargs commented:
Fair enough, re-added. Assuming tests pass, I believe I've taken care of all the outstanding items.

cjcenizal commented on src/core_plugins/kibana/public/dashboard/dashboard.html:
```diff
@@ -48,7 +47,7 @@
             </button>
           </div>
-          <kbn-typeahead-items></kbn-typeahead-items>
+          <kbn-typeahead-items><query-help></query-help></kbn-typeahead-items>
```
I internally debated over asking for this because it's kind of nitpicky, but I might as well... can we format this to use indentation to reflect the nested natured of these elements?

```html
<kbn-typeahead-items>
  <query-help></query-help>
</kbn-typeahead-items>
```

Bargs replied:
Certainly, not a problem. Updated!

cjcenizal commented on src/core_plugins/kibana/public/discover/index.html:
```diff
@@ -45,7 +44,7 @@
             </button>
           </div>
-          <kbn-typeahead-items></kbn-typeahead-items>
+          <kbn-typeahead-items><query-help></query-help></kbn-typeahead-items>
```
Same here.

cjcenizal commented on src/core_plugins/kibana/public/visualize/editor/editor.html:
```diff
@@ -59,7 +58,7 @@
               </button>
             </div>
-            <kbn-typeahead-items></kbn-typeahead-items>
+            <kbn-typeahead-items><query-help></query-help></kbn-typeahead-items>
```
Same here.

Bargs commented:
@lukasolson mind giving this a quick look as well?

lukasolson commented on src/core_plugins/kibana/public/dashboard/dashboard.html:
```diff
@@ -28,7 +28,6 @@
           <div class="kuiLocalSearch">
             <input
               parse-query
-              input-focus
```
Hmm, I'm not a huge fan of this. Now, when Discover loads, I can't get to work immediately without using my mouse to focus in on the search input.

What if, instead, we put a help icon to the right side of the search, and when you clicked on it, it brought up this same help dialog?

/cc @cjcenizal

lukasolson replied:
Ultimately, if/when https://github.com/elastic/kibana/pull/10976 gets in, we could move to something more like how Google Drive does this:

![image](https://cloud.githubusercontent.com/assets/1178348/24634357/62a2e32a-1882-11e7-92c3-17dd192e88c2.png)

Notice how there's a "Learn More" button in the bottom of the pop-up. Thoughts?

Bargs replied:
I don't think removing `input-focus` is actually necessary to accomplish the current experience, I just removed it when I was playing around with things. I left it out because I remember this auto-focus came up as something that hurts accessibility in the accessibility study that was done awhile back. If we want to discuss that outside of this PR, I can try putting it back in.

cjcenizal replied:
If auto-focus is a critical usability improvement, then I think we should keep it and figure out other ways to address the accessibility concerns, e.g. providing context to screen readers with aria attributes.

FYI @snide is working on some designs this week that incorporate both these changes and the ones in https://github.com/elastic/kibana/pull/10976.

lukasolson replied:
I think ultimately if we do show this help every time the user has focus on the search bar, we should at least have a mechanism to hide it, because I can see it getting in the way, and potentially frustrating for users if they don't want to see the help any more.

Bargs replied:
To do that I think we'll need to store the user's choice in local or session storage, which we don't have a great way to do in a consistent manner in Kibana at the moment. 

The only issues I have with putting a help link in the dropdown added by https://github.com/elastic/kibana/pull/10976 is that 1. it will be kind of buried and 2. this help text doesn't relate to the rest of the content of the dropdown. The dropdown manages filters while this help text is about the query bar.

It sounds like the help should be hidden by default, but easy to find. Should we put a help link in or near the query bar? 

Or, we could go back to the original implementation that hid the help after the user starts typing. I think there was significantly less risk of that becoming annoying since it wasn't in the user's face any time the query bar had focus.

lukasolson replied:
Fair enough. Let's just stick with the original implementation for now.

lukasolson commented:
Hmm, looks like the query help is sticking around for me even after I start typing.

Bargs commented:
@lukasolson yes, it was changed to be persistent at @cjcenizal's request. I can revert it if he's ok with that solution.

tbragin commented:
I tried it out, and I think I prefer the persistent behavior. 

@bargs this is really awesome! With feature freeze having passed for 5.4, I assume this will have to go into 5.5?

Bargs commented:
@tbragin @lukasolson @cjcenizal @snide 

New update based on our convo:

![screen shot 2017-04-12 at 1 12 26 pm](https://cloud.githubusercontent.com/assets/6239176/24970497/88a6fba4-1f82-11e7-9965-05db953594c4.png)

I have to say I quite like it, simple and gets the job done. Let me know what you all think. 

A couple notes:
* If you pull down the PR, the doc link will be a 404 at the moment because there are no 6.0 docs. The link changes dynamically based on the version of Kibana you're running. Just switch `6.0` to `current` in the url to see which page it's pointing at.
* I've only updated the discover page so far, if we can reach agreement on the design I'll go ahead and update the other pages.

alexfrancoeur commented:
@Bargs what are your thoughts on adding a "don't show me this again" option for power users? I could see this being a bit annoying if you are comfortable with lucense

This is a bit minor but any thoughts on how previous queries and this in-line help could play better together?
<img width="1255" alt="screen shot 2017-04-12 at 1 00 10 pm" src="https://cloud.githubusercontent.com/assets/25181823/24969941/8c26e516-1f80-11e7-9740-c7005773addc.png">

Bargs commented:
Thanks for the reminder @cjcenizal, I didn't know about those files.

@alexfrancoeur I removed that inline help, the newest version just provides a link to the docs in the query bar.

Bargs commented on ui_framework/components/local_nav/_local_search.scss:
```diff
@@ -4,47 +4,50 @@
-    border-color: $localSearchBackgroundColor--darkTheme;
+  .kuiLocalSearch__input {
+    flex: 1 1 100%;
+    padding: 5px 13em 5px 28px;
```
@cjcenizal not sure if this line is what prompted you to ask about `em`s, but in case you were wondering I used `em`s for the right padding here so that it would still be correct even if the user changes their browser font size. Otherwise the doc link and the query string could overlap. I can add a comment to that effect if you guys think it's confusing.

cjcenizal replied:
Great reasoning. Thanks for explaining. Could you add a comment everywhere you used `em`? Our format for CSS comments is:

```scss
/**
 * 1. Comment one, which applies to multiple properties.
 * 2. Comment two.
 */
.class {
  display: block; /* 1 */
  color: red; /* 2 */
  width: 100px; /* 1 */
}
```

Bargs commented:
@tbragin @snide @lukasolson still looking for feedback on the latest iteration: https://github.com/elastic/kibana/pull/10928#issuecomment-293648861

cjcenizal commented on src/core_plugins/kibana/public/discover/index.html:
```diff
@@ -25,25 +25,19 @@
               kbn-typeahead-input
               ng-model="state.query"
-              placeholder="Search..."
+              placeholder="Search... ex. status:200 AND extension:PHP"
```
Can we use "e.g." instead of "ex."? I think this is more common, and it's also used in our writing style guide.

Bargs replied:
👍  https://github.com/elastic/kibana/pull/10928/commits/f120beaf19aed18333025a2c9961d257777202ff

cjcenizal commented on src/core_plugins/kibana/public/discover/index.html:
```diff
@@ -25,25 +25,19 @@
-            >
-              <span class="fa fa-search" aria-hidden="true"></span>
-            </button>
+            <a class="kuiLink kuiLocalSearch__help" ng-href="{{queryDocLinks.luceneQuerySyntax}}" target="_blank">Uses lucene query syntax</a>
```
According to our HTML style guide, we can improve readability by breaking this up into multiple lines:

```html
<a
  class="kuiLink kuiLocalSearch__help"
  ng-href="{{queryDocLinks.luceneQuerySyntax}}"
  target="_blank"
>
  Uses lucene query syntax
</a>
```

Bargs replied:
👍  https://github.com/elastic/kibana/pull/10928/commits/f120beaf19aed18333025a2c9961d257777202ff

alexfrancoeur commented:
@Bargs I like it. Simple and gets the job done. It looks fine if there is a long query in there as well.

<img width="578" alt="screen shot 2017-04-19 at 10 12 21 am" src="https://cloud.githubusercontent.com/assets/25181823/25184532/c3836904-24e8-11e7-9a18-a9572eeff9b7.png">

I know you mentioned the link should be a 404 but clicking on the link currently doesn't do anything for me. I assume this will open a new tab and then focus on that tab - is that correct? Or are we opening a new window?

Bargs commented:
@alexfrancoeur that's odd, you should definitely be getting a new tab that takes you to the URL that will eventually hold the 6.0 docs. What browser are you using?

Bargs commented:
@alexfrancoeur it appears there was a conflict with another PR that was merged into master, link should be fixed now.

Bargs commented:
@cjcenizal I addressed your feedback. I also had to handle some merge conflicts with the changes in https://github.com/elastic/kibana/pull/11287 so you might wanna take a look and make sure I didn't screw anything up. I undid the changes to the class names, so I don't think the framework examples need updated anymore?

cjcenizal commented:
While testing this out I noticed some problems with the search in Timelion:

![image](https://cloud.githubusercontent.com/assets/1238659/25230693/ccf0cb84-2589-11e7-9718-f974775038b4.png)

I think we need to create some general-use components to give us more fine-grained control over the search. I created https://github.com/elastic/kibana/pull/11343 for this purpose. @Bargs Could you take a look and let me know if this will work for you?

alexfrancoeur commented:
@Bargs looks good now - thanks!

cjcenizal commented:
We just need to rebase and update the markup to use the new components, and we should be good to 🐐!

lukasolson commented on src/core_plugins/kibana/public/discover/index.html:
```diff
@@ -30,12 +30,19 @@
               kbn-typeahead-input
               ng-model="state.query"
-              placeholder="Search..."
+              placeholder="Search... e.g. status:200 AND extension:PHP"
```
Minor feedback, take it or leave it: I think it might help to throw in parentheses here so it's "Search... (e.g. status:200 AND extension:PHP)" but that's just my preference.

tbragin commented:
@bargs Checked it out, and here is my reaction:
* I think something is lost by not having the drop-down right in front of me with common queries when I type. I can open a separate page with docs, but at that point I'm switching back and forth. However, I don't want to fight that battle now... even making the docs easier to find is an improvement.
* I agree with @lukasolson that we need the same link to help from the "no results screen" and this reference. 
* Do we need "Uses Lucene query syntax" *always* there? I can see how that can get annoying in the long run. This phrase is going to be in *every* screenshot of Kibana Discover going forward, think about it. Can we just have a little (i) icon or something with a mouseover that pulls up the query help?
* Why is the behavior different for search bar in Discover vs Visualize and Dashboard?

<img width="1422" alt="screen shot 2017-05-03 at 7 50 37 am" src="https://cloud.githubusercontent.com/assets/933397/25666373/3cd575e4-2fd5-11e7-98a3-11052584f31b.png">

<img width="1418" alt="screen shot 2017-05-03 at 7 50 27 am" src="https://cloud.githubusercontent.com/assets/933397/25666371/3a1053ba-2fd5-11e7-96d1-fefa25bef4c7.png">

Bargs commented:
Thanks for the feedback @tbragin 

> we need the same link to help from the "no results screen" and this reference.

Agreed, I'll make that update.

> Why is the behavior different for search bar in Discover vs Visualize and Dashboard?

I was just waiting for consensus on design before updating all of the other apps.

> Do we need "Uses Lucene query syntax" always there? I can see how that can get annoying in the long run. This phrase is going to be in every screenshot of Kibana Discover going forward, think about it. Can we just have a little (i) icon or something with a mouseover that pulls up the query help?

It sounds like this is the last bit we need to iron out. @snide @cjcenizal what do you think? I don't personally have a strong opinion.

cjcenizal commented:
It might be nice to establish a common icon for helping people identify links to helpful resources, but this is definitely a @snide decision.

snide commented:
@tbragin 

> Do we need "Uses Lucene query syntax" always there? I can see how that can get annoying in the long run. This phrase is going to be in every screenshot of Kibana Discover going forward, think about it. Can we just have a little (i) icon or something with a mouseover that pulls up the query help?

You can always make this appear on focus of the input so that it doesn't show until it matters. If you're targeting BI users even partially, id suggest keeping that element there, otherwise people will just assume it's a general google search bar.

Id prefer we stop using the "i" buttons when we can. More specifically here because it needs to link to the docs.

Bargs commented:
@lukasolson @cjcenizal just pushed some updates. I believe I've covered everything we've talked about. Ready for another review.

* rebase, use new components
* update placeholder with parenthesis
* make help link show up conditionally on input focus
* update no results page help link
* Update dashboard and visualize pages

Bargs commented:
jenkins, test this

cjcenizal commented on ui_framework/components/local_nav/_local_search.scss:
```diff
@@ -28,6 +28,10 @@
 
+.kuiLocalLuceneSearchInput:focus {
+  padding-right: 13em;
+}
```
I'd characterize this as a custom class, specific to the way we're using the LocalSearch components in Discover and elsewhere in Kibana. Looking at it from that perspective, I think it's probably best to rename this to something like `localLuceneSearchInput` (reserving the kui prefix for the UI Framework classes), and put it in `src/core_plugins/kibana/public/kibana.less` (or somewhere within that directory).

I think the counter-point is that we want to use this help link in every instance of the LocalSearch, but I think we should still draw a line and use the UI Framework to provide generic components and guidance on patterns for using them, and use `src` to define more concrete rules on how these generic components are used in Kibana. What do you think @snide ?

snide replied:
Or just more generically name it as a modifier, which can let it still live here and be reused (since it seems pretty generic).

cjcenizal replied:
Normally I'd agree, but the 13em padding is really tightly coupled with the length of the string that goes in there, i.e. "Uses Lucene query syntax" is 13em wide.

cjcenizal replied:
But it's no biggie, we can keep it here as a modifier. It'd be easy to move if it becomes confusing.

Bargs replied:
Will other plugins be able to use this class if we put it in `src/core_plugins/kibana/public/kibana.less`? If yes, then I think that's fine. If no, it would be nice to keep it somewhere reusable. 

By name it as a modifier, do you mean something like `kuiLocalSearchInput--lucene`?

cjcenizal replied:
If someone disables the Kibana plugin then I guess those styles might not be available. If that's a concern, then how about putting them into `ui/public/styles/local_search.less`?

Yep, that uses the correct modifier format. Looks good to me.

cjcenizal replied:
Maybe add a comment to the styles noting that `13em is the width of 'Uses Lucene query syntax'` so it isn't just a magic number.

Bargs replied:
`ui/public/styles/local_search.less` sounds fine to me. Just to be clear, which name do you prefer in that file, `localLuceneSearchInput` or `localSearchInput-lucene`? Or something else?

cjcenizal replied:
Let's use the modifier format to indicate how it's related to the base class: `kuiLocalSearchInput--lucene`.

Bargs replied:
https://github.com/elastic/kibana/pull/10928/commits/30b86221729d233d716788e1f187076d96746b0d

cjcenizal commented on src/core_plugins/kibana/public/discover/partials/no_results.html:
```diff
@@ -26,7 +26,7 @@
     <h3>Refine your query</h3>
     <p>
-      The search bar at the top uses Elasticsearch's support for Lucene <a href="http://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html#query-string-syntax" target="_blank">Query String syntax</a>. Let's say we're searching web server logs that have been parsed into a few fields.
+      The search bar at the top uses Elasticsearch's support for Lucene <a ng-href="{{queryDocLinks.luceneQuerySyntax}}" target="_blank">Query String syntax</a>. Let's say we're searching web server logs that have been parsed into a few fields.
```
Can you add a `kuiLink` class here?

Bargs replied:
👍  https://github.com/elastic/kibana/pull/10928/commits/8d8818e4097f437647e3293a2bad764dd1729a78

cjcenizal commented on ui_framework/components/local_nav/_local_search.scss:
```diff
@@ -44,6 +48,15 @@
+  .kuiLocalSearchAssistedInput__assistanceLink:active {
+    display: inherit;
+  }
+
```
Can we replace lines 31 through 69 with this:

```scss
/**
 * 1. Visually hide this until the parent is focused, so that it's accessible to screen readers.
 */
.kuiLocalSearchInput:focus {
  & ~ .kuiLocalSearchAssistedInput__assistance {
    .kuiLocalSearchAssistedInput__assistanceLink {
      opacity: 1; /* 1 */
    }
  }
}

.kuiLocalSearchAssistedInput {
  display: flex;
  flex: 1 1 100%;
  position: relative;
}

  /**
   * 1. Vertically center the assistance, regardless of its height.
   */
  .kuiLocalSearchAssistedInput__assistance {
    position: absolute;
    right: $globalFormControlHorizontalPadding;
    top: 50%; /* 1 */
    transform: translateY(-50%); /* 1 */
  }

  /**
   * 1. Visually hide this until it's focused, so that it's accessible to screen readers, and is
   *    also keyboard-accessible.
   */
  .kuiLocalSearchAssistedInput__assistanceLink {
    opacity: 0; /* 1 */

    &:active,
    &:focus {
      opacity: 1; /* 1 */
    }
  }
```

These changes will ensure the link is visible to screen readers, and can also be tabbed to by users who are using the keyboard to navigate.

Bargs replied:
The issue with this approach is that the link will now interfere with interactions with the input:

![opacity](https://cloud.githubusercontent.com/assets/6239176/25716631/b0d6af64-30cd-11e7-9fcd-abc78d6129f9.gif)

cjcenizal replied:
Good point. How about if we change the help to show up on hover instead of focus:

```scss
.kuiLocalSearchInput:hover {
  & ~ .kuiLocalSearchAssistedInput__assistance {
    .kuiLocalSearchAssistedInput__assistanceLink {
      opacity: 1; /* 1 */
    }
  }
}
```

Bargs replied:
That seems like a weird interaction to me. If the user wants to modify something at the end of their query they're going to mouse over it in order to click the spot they want to edit, but then that spot is going to suddenly disappear under the help link. The only way for them to get to what they wanted at that point is to click another part of the query and shift the cursor with the keyboard. I'd find that infuriating as a user.

cjcenizal replied:
It seems like we have that problem anyway:

![lucene_help_link](https://cloud.githubusercontent.com/assets/1238659/25717822/9b1ba4be-30b8-11e7-98f3-1878ebf956f8.gif)

As long as we hide that link by default, it's going to create a strange interaction when users try to click the text that it invisibly occupies. Maybe we should go back to showing it all the time? @snide what do you think?

Bargs replied:
You're right, that interaction stinks as well. Unless @snide has any good ideas, tomorrow I'll switch it back to showing all the time so this doesn't become a big time sink.

Bargs replied:
https://github.com/elastic/kibana/pull/10928/commits/30b86221729d233d716788e1f187076d96746b0d

cjcenizal commented on ui_framework/dist/ui_framework.css:
```diff
@@ -706,6 +706,7 @@ body {
   /* 2 */
   font-size: 10px !important;
   /* 2 */
+  -webkit-transition: background-color 0.1s linear;
```
There's something funny going on where some people have been generating CSS that adds these properties, and yet I've been generating CSS that removes them. Are you using yarn to install your NPM deps by any chance? Which version of postcss do you have installed? I have 5.2.17.

Bargs replied:
Not using yarn but I was on 5.2.11. I installed fresh, now I have 5.2.17 and I'm generating the same css as you (with the lines removed).

cjcenizal replied:
😅  Thank you!

Bargs commented:
@cjcenizal @lukasolson alright, ready for another look.

* Reverted show/hide on focus/unfocus behaviour
* Moved `.kuiLocalSearchInput--lucene` to `ui/public/styles/local_search.less`
* Added `kuiLink` class to no results page doc link
